sklearn/discriminant_analysis.py

```diff
@@ -25,47 +25,63 @@
 from .utils.extmath import softmax
 from .preprocessing import StandardScaler
 
-
 __all__ = ['LinearDiscriminantAnalysis', 'QuadraticDiscriminantAnalysis']
 
 
-def _cov(X, shrinkage=None):
-    """Estimate covariance matrix (using optional shrinkage).
-
+def _cov(X, shrinkage=None, covariance_estimator=None, standardize=False):
+    """Estimate covariance matrix (using covariance_estimator).
     Parameters
     ----------
     X : array-like, shape (n_samples, n_features)
         Input data.
-
+    covariance_estimator : BaseEstimator
+        A covariance estimator, possible values
+            - None: shrinkage parameter drives the estimate
+            - BaseEstimator with a fit method and
+            a .covariance_ attribute like the estimators
+            in sklearn.covariance
+    standardize : bool
+        if True standardize the data before applying the covariance estimator
+        Note: only used when covariance_estimator is not None
     shrinkage : string or float, optional
         Shrinkage parameter, possible values:
           - None or 'empirical': no shrinkage (default).
           - 'auto': automatic shrinkage using the Ledoit-Wolf lemma.
           - float between 0 and 1: fixed shrinkage parameter.
-
     Returns
     -------
     s : array, shape (n_features, n_features)
         Estimated covariance matrix.
     """
-    shrinkage = "empirical" if shrinkage is None else shrinkage
-    if isinstance(shrinkage, str):
-        if shrinkage == 'auto':
-            sc = StandardScaler()  # standardize features
+    if covariance_estimator is None:
+        shrinkage = "empirical" if shrinkage is None else shrinkage
+        if isinstance(shrinkage, str):
+            if shrinkage == 'auto':
+                sc = StandardScaler()  # standardize features
+                X = sc.fit_transform(X)
+                s = ledoit_wolf(X)[0]
+                # rescale
+                s = sc.scale_[:, np.newaxis] * s * sc.scale_[np.newaxis, :]
+            elif shrinkage == 'empirical':
+                s = empirical_covariance(X)
+            else:
+                raise ValueError('unknown shrinkage parameter')
+        elif isinstance(shrinkage, float) or isinstance(shrinkage, int):
+            if shrinkage < 0 or shrinkage > 1:
+                raise ValueError('shrinkage parameter must be between 0 and 1')
+            s = shrunk_covariance(empirical_covariance(X), shrinkage)
+        else:
+            raise TypeError('shrinkage must be of string or int type')
+    else:
+        if standardize:
+            sc = StandardScaler()
             X = sc.fit_transform(X)
-            s = ledoit_wolf(X)[0]
-            # rescale
+            covariance_estimator.fit(X)
+            s = covariance_estimator.covariance_
```

Comment:
It's probably a good idea to do some validation here and check if the fitted estimator actually has `covariance_` set. If not, raise the appropriate RuntimeError or something with a helpful message.